File: Top-Interview-Questions-Leetcode/Easy-Collections/design/shuffle-an-array.cpp
#include <iostream>
#include <vector>
#include <random>

class Solution {
private:
    
    std::vector<int> origin;
    std::mt19937 gen;
    
public:
    Solution(std::vector<int>& nums) : origin(nums), gen(std::random_device()()) {}
    
    std::vector<int> reset() {
        return this->origin;
    }
    
    std::vector<int> shuffle() {
        int sz = this->origin.size();
        std::uniform_int_distribution<> distrib;

        std::vector<int> nums(origin.begin(), origin.end());

        for (int i = sz - 1, idx; i >= 0; --i) {
            distrib.param(std::uniform_int_distribution<>::param_type(0, i));
            idx = distrib(gen);
            std::swap(nums[i], nums[idx]);
        }

        return nums;
    }
    
    // Alternative implementation using rand()
    std::vector<int> shuffle_rand() {
        int sz = this->origin.size();
        std::vector<int> nums = origin;
        
        for (int i = sz - 1, idx; i >= 0; --i) {
            idx = rand() % (i + 1);
            std::swap(nums[idx], nums[i]);
        }
        
        return nums;
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(nums);
 * vector<int> param_1 = obj->reset();
 * vector<int> param_2 = obj->shuffle();
 */

File: Top-Interview-Questions-Leetcode/Easy-Collections/design/shuffle-an-array.py
from typing import List
import random

class Solution:

    def __init__(self, nums: List[int]):
        self.origin = nums[:]

    def reset(self) -> List[int]:
        return self.origin

    def shuffle(self) -> List[int]:
        nums = self.origin[:]
        
        for i in range(len(nums) - 1, -1, -1):
            idx = random.randint(0, i)
            nums[idx], nums[i] = nums[i], nums[idx]
            
        return nums


# Your Solution object will be instantiated and called as such:
# obj = Solution(nums)
# param_1 = obj.reset()
# param_2 = obj.shuffle()